cpp/include/cudf/utilities/span.hpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <limits>
+#include <memory>
+#include <type_traits>
+
+namespace cudf {
+namespace detail {
+
+constexpr size_t dynamic_extent = std::numeric_limits<std::size_t>::max();
+
+/**
+ * @brief C++20 std::span with reduced feature set.
+ */
+template <typename T, size_t Extent = dynamic_extent>
+class span {
```

Comment:
I think this class should be called `device_span` to indicate that the viewed memory resides on the device.

Reply:
`span` is designed to be non-owning. Is it important to make the distinction between host and device? There are use cases where we would want to implicitly convert from `span` -> `device_span`, at least.

Reply:
> span is designed to be non-owning. Is it important to make the distinction between host and device?

It doesn't really have anything to do with ownership. If you attempt to access host memory on device or device memory on host, that is a hard error. Standard C++ doesn't have separate memory spaces like this, so `std::span` doesn't have to worry about it. We do. This is why `thrust` has both `host_vector` and `device_vector`. The accessibility of the memory being viewed needs to be somehow encoded in the type.

Reply:
So strictly speaking there's no reason we _can't_ use just one `span` class... but the intent is more explicit.

Reply:
We'd want both `span` and `device_span` then, where `device_span` is implicitly convertible, I suppose?

Reply:
Is the intent to call the methods of this class from `__device__` code? Or just to pass its pointer and extent to `__device__` code?

Reply:
Call the functions. I haven't tests for device code yet, but it's now glaringly obvious that I've omitted `__device__`.

Reply:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#constexpr-functions

If we compile with `--expt-relaxed-constexpr` you may not need to, if all the methods are `constexpr`...

Reply:
It would appear that we do. That would be convenient.

https://github.com/rapidsai/cudf/blob/b35882895513143bb05e7906c525f9f83788fea9/cpp/CMakeLists.txt#L119

If that's the case, we could us the CRTP (curiously recurring template pattern) to avoid having two classes, and just base them both off of an abstract `span_base` class.

Reply:
> where device_span is implicitly convertible, I suppose?

I do not think they should be implicitly convertible. That would be too error prone.

Reply:
Yeah I've backtracked on that myself after writing some tests. Doesn't make sense to convert between the two. I originally misunderstood the purposes between the two classes.